index.js

```
@@ -1,219 +1,160 @@
-			width:   12,
-			default: '127.0.0.1',
-			regex:   self.REGEX_IP
-		},{
-			type:    'textinput',
-			id:      'port',
-			label:   'Port number',
-			width:   6,
-			default: '2000',
-			regex:   self.REGEX_PORT
-		}
-	]
-};
 
-// When module gets deleted
-instance.prototype.destroy = function() {
-	var self = this;
+	incomingData(data) {
+		this.log('debug', data)
 
-	if (self.socket !== undefined) {
-		self.socket.destroy();
+		this.updateStatus(InstanceStatus.Ok, 'Received data')
+		this.log('info', 'logged in')
 	}
 
-	debug("destroy", self.id);
-};
-
-instance.prototype.actions = function(system) {
-	var self = this;
-	self.setActions(actions.getActions());
-};
+	async init(config) {
+		var self = this
 
-instance.prototype.initPresets = function (updates) {
-	var self = this;
-	self.setPresetDefinitions(presets.getPresets(self));
-};
+		this.updateStatus(InstanceStatus.Disconnected)
 
-instance.prototype.action = function(action) {
+		this.config = config
 
-	var self = this;
-	var id = action.action;
-	var cmd;
+		debug = self.debug
+		log = self.log
 
-	switch (id) {
-		case 'go':
-			cmd = 'go';
-			break;
-
-		case 'stop_all':
-			cmd = 'stop all';
-			break;
-
-		case 'fade_all':
-			cmd = 'fade all';
-			break;
-
-		case 'pause_all':
-			cmd = 'pause all';
-			break;
-
-		case 'resume_all':
-			cmd = 'resume all';
-			break;
+		this.init_tcp()
+		self.initPresets()
+		actions(this)
 
-		case 'stopwatch_stop':
-			cmd = 'stopwatch stop';
-			break;
-
-		case 'stopwatch_start':
-			cmd = 'stopwatch start';
-			break;
+		//		self.setActionDefinitions(actions.getActions()); // export actions
+	}
 
-		case 'stopwatch_reset':
-			cmd = 'stopwatch reset';
-			break;
+	init_tcp() {
+		var self = this
 
-		case 'advance':
-			cmd = 'advance';
-			break;
+		if (self.socket !== undefined) {
+			self.socket.destroy()
+			delete self.socket
+			self.login = false
+		}
 
-		case 'pause':
-			cmd = 'pause';
-			break;
+		if (self.config.host) {
```

should check for existance **and** validiy of host **and** port

---

What exactly do you mean, won't the Regex handle that?

---

Yes and no. In actions or feedbacks the regex is just a visual hint, in the configuration you can't hit save if the regex doesn't match. This check in the config has not been there from the start so there may be configurations with illegal data. Also some users edit the configuration database by hand. It is always a good idea to do some extra checks. Minimum goal is that the module doesn't crash if it gets invalid data. You can go as complicated as you want.

---

So it turns out I'd not changed from `this.REGEX_IP` to `Regex.IP` and it was silently dropping all regex checks in the config, without throwing any errors during build, so that was a good fix at least.

Can you point me at another module that does this I could learn from, I've got something that actually works (but haven't committed it), after the obvious thing didn't work because a regex string needs `foo` whereas a regex literal needs `/foo/` in JS!

Again it feels like we want to nick a chunk of:
https://github.com/bitfocus/companion/blob/master/webui/src/Components/TextInputField.jsx#L56-L66

Into companion-module-base if we're going to do this, and fetch the config from getConfigFields, so I just pass in the field name, and maybe the variable, and it gives me a boolean result, again rather than everyone reinventing the wheel badly/with different edge cases.